thunder/core/update_aliases.py

```diff
@@ -0,0 +1,107 @@
+from collections import defaultdict
+from itertools import chain
+from functools import reduce, partial
+
+import thunder.core.prims as prims
+from thunder.core.proxies import TensorProxy, variableify, unvariableify
+from thunder.core.pytree import tree_flatten, tree_unflatten
+from thunder.core.symbol import BoundSymbol
+from thunder.core.trace import from_trace, tracectx, TraceCtx as Trace, TraceProvenance
+
+
+def _get_update_bsym(group, swap_map, new_aliases):
+    aliases = tuple(unvariableify(alias) for alias in group)
+    update_aliases_bsym = prims.update_aliases.bind(aliases, output=new_aliases)
+    update_aliases_bsym = update_aliases_bsym.from_bsym_swap_proxies(swap_map)
+    for new_alias, old_alias in zip(new_aliases, group):
+        swap_map[old_alias] = new_alias
+    return update_aliases_bsym, swap_map
+
+
+def _get_new_alias(alias, trace):
+    with tracectx(trace):
+        # !! better to use the meta function?
+        new_alias = TensorProxy(like=alias.proxy, requires_grad=alias.proxy.requires_grad)
+    return new_alias
+
+
+def _get_new_aliases(aliases, trace):
+    return tuple(map(partial(_get_new_alias, trace=trace), aliases))
+
+
+def _is_inplace_op(bsym):
+    return bsym.sym.tags and prims.OpTags.IN_PLACE in bsym.sym.tags
+
+
+def _is_view_creation_op(bsym):
+    import thunder.torch as ltorch
+
+    # !!! check on _syms_that_may_return_views?
+    return bsym.sym in ltorch._syms_returning_views
+
+
+def _involves_viewed_args(bsym, viewed):
+    if bsym.sym.id == prims.PrimIDs.RETURN:
+        return False
+    return any(isinstance(p, TensorProxy) and variableify(p) in viewed for p in bsym.flat_proxy_args)
+
+
+def insert_alias_updates(computation_trace: Trace) -> list[Trace]:
+    swap_map = dict()
+    bsyms = []
+
+    # First pass: identify views, their originals, and operands involved in inplace ops
+    view_groups = []
+    inplace_inputs = set()
+    for bsym in computation_trace.bound_symbols:
+        if _is_inplace_op(bsym) or _is_view_creation_op(bsym):
+            # only interested in the input which is modified by the inplace op
+            in_tensor = list(map(variableify, filter(lambda p: isinstance(p, TensorProxy), bsym.flat_proxy_args)))[0]
```

Comment:
In my gut feeling for most ops this would be equivalent to `variableify(bsym.flat_proxy_args[0]`. Correct me if I'm wrong.
I guess this could be simpler if we have a map from op to index of tensor arg to be modified.
What do you think?

Reply:
Yes, you are right about the equivalence to `variableify(bsym.flat_proxy_args[0]`.

This does assume that the first argument to an in-place op or a view creation op is the thing needing to be aliased.  If this is the case, then I think what's done here is simplest.  If it is not the case, then having such a map would be the way to go.  I think having such a map pushes the complexity to another place, and makes adding operators more difficult.

Reply:
> This does assume that the first argument to an in-place op or a view creation op is the thing needing to be aliased. If this is the case, then I think what's done here is simplest.

Agreed. Okay let's keep it.

some clarification: The reason I asked is that we already have a map which is tad similar to what I had in my mind -- https://github.com/Lightning-AI/lightning-thunder/blob/92a6adb63ddcf36e3910a55fc5f7a02cec5bee8c/thunder/torch/__init__.py#L93-L95. So I thought this line could be written as e.g. `variableify(bsym.flat_proxy_args[some_map.get(bsym.sym, 0)])`